src/components/CadenceEditor/index.tsx

```diff
@@ -0,0 +1,189 @@
+import { EditorContainer } from './components';
+
+import { EditorState } from './types';
+
+const MONACO_CONTAINER_ID = 'monaco-container';
+
+const CadenceEditor = (props: any) => {
+  const project = useProject();
+  const [editor, setEditor] = useState(null);
+  const editorOnChange = useRef(null);
+  // We will specify type as index as non-existent numbers to prevent collision with existing enums
+  const lastEdit = useRef({
+    type: 8,
+    index: 8,
+  });
+
+  const [editorStates, setEditorStates] = useState({});
+
+  // TODO: restore view state in next implementation
+  /*
+  const saveEditorState = (id: string, viewState: any) => {
+    setEditorStates({
+      ...editorStates,
+      [id]: viewState,
+    });
+  };
+   */
+
+  // This method is used to retrieve previous MonacoEditor state
+  const getOrCreateEditorState = (id: string, code: string): EditorState => {
+    const existingState = editorStates[id];
+
+    if (existingState !== undefined) {
+      return existingState;
+    }
+
+    const newState = {
+      model: monaco.editor.createModel(code, CADENCE_LANGUAGE_ID),
+      viewState: null,
+    };
+
+    setEditorStates({
+      ...editorStates,
+      [id]: newState,
+    });
+
+    return newState;
+  };
+
+  // "getActiveCode" is used to read Cadence code from active(selected) item
+  const getActiveCode = () => {
+    const { active } = project;
+    const { accounts, scriptTemplates, transactionTemplates } = project.project;
+
+    const { type, index } = active;
+    let code, id;
+    switch (type) {
+      case EntityType.Account:
+        code = accounts[index].draftCode;
+        id = accounts[index].id;
+        break;
+      case EntityType.TransactionTemplate:
+        code = transactionTemplates[index].script;
+        id = transactionTemplates[index].id;
+        break;
+      case EntityType.ScriptTemplate:
+        code = scriptTemplates[index].script;
+        id = scriptTemplates[index].id;
+        break;
+      default:
+        code = '';
+        id = 8;
+    }
+    return [code, id];
+  };
+
+  // Method to use, when model was changed
+  const debouncedModelChange = debounce(() => {
+    // we will ignore text line, cause onChange is different for readme and other scripts
+    // @ts-ignore
+    project.active.onChange(editor.getValue());
+  }, 150);
+
+  useEffect(configureCadence, []);
+  useEffect(() => {
+    // TODO: save/restore view state with:
+    //  - use ref to track current active id
+    //  - const oldState = editor.saveViewState();
+    if (editor) {
+      // Remove tracking of model updates to prevent re-rendering
+      if (editorOnChange.current) {
+        editorOnChange.current.dispose();
+      }
+
+      // To pick up new changes in accounts, we will track project's active item and then add and remove
+      // new line at EOF to trick Language Client to send code changes and reimport the latest changes,
+      // clearing errors and warning about missing fields.
```

Comment:
Interesting hack. What if there is no newline to strip?

Reply:
There always will be one - cause we are adding it before removing :)